### PR TITLE
Do not bind mysql parameters if there are none

### PR DESCRIPTION
**Pull request by ChillerDragon**
Continuation of https://github.com/ddnet/ddnet/pull/9131

### CODE REVIEW

**Review by def- — APPROVED** [review]

**Comment by ChillerDragon** [comment]
Fixed conflicts

**Comment by ChillerDragon** [comment]
Flaky test?

``[  FAILED  ] Net.Ipv4AndIpv6Work``

**Comment by heinrich5991** [comment]
What is the motivation for this PR? Does the function fail if there are no bound parameters? Please include such motivation in the commit description.

File: src/engine/server/databases/mysql.cpp [hunk 1]
```diff
@@ -446,7 +446,8 @@ bool CMysqlConnection::Step(bool *pEnd, char *pError, int ErrorSize)
 	if(m_NewQuery)
 	{
 		m_NewQuery = false;
-		if(mysql_stmt_bind_param(m_pStmt.get(), m_vStmtParameters.data()))
+		unsigned NumParameters = mysql_stmt_param_count(m_pStmt.get());
+		if(NumParameters && mysql_stmt_bind_param(m_pStmt.get(), m_vStmtParameters.data()))
```
Comment (heinrich5991) [inline src/engine/server/databases/mysql.cpp:450]:
```suggestion
		if(m_vStmtParameters.size() != 0 && mysql_stmt_bind_param(m_pStmt.get(), m_vStmtParameters.data()))
```

Comment (heinrich5991) [reply src/engine/server/databases/mysql.cpp:450]: Oh, there's even [`std::vector::empty`](https://en.cppreference.com/w/cpp/container/vector/empty).

**Comment by ChillerDragon** [comment]
> What is the motivation for this PR? Does the function fail if there are no bound parameters? Please include such motivation in the commit description.

Same as #9131 the code throws an asan error if you run a sql query without parameters.

```
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_vector.h:1124:9: runtime error: reference binding to null pointer of type 'st_mysql_bind'
    #0 0x55555604931c in std::vector<st_mysql_bind, std::allocator<st_mysql_bind> >::operator[](unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_vector.h:1124:2
    #1 0x555556043414 in CMysqlConnection::ExecuteUpdate(int*, char*, int) /home/chiller/git/ddnet-insta/src/engine/server/databases/mysql.cpp:480:44
 ```

File: src/engine/server/databases/mysql.cpp (hunk 1 above)
Comment (ChillerDragon) [reply src/engine/server/databases/mysql.cpp:450]: updated

**Comment by heinrich5991** [comment]
The ASAN issue you posted is from before #9131. Does the new code also have such a problem? If not, this shouldn't be merged.

**Comment by ChillerDragon** [comment]
https://github.com/ddnet/ddnet/pull/9131 is what unlocked this new asan error it used to crash earlier now it crashes here. The square brackets were removed in https://github.com/ddnet/ddnet/pull/9130 which should only be readability and not fix the null pointer issue as far as I understand it.

**Comment by heinrich5991** [comment]
> The square brackets were removed in #9130 which should only be readability

https://clang.llvm.org/extra/clang-tidy/checks/readability/container-data-pointer.html states it's not only about readability, but actually UB when the container is empty:

> This also ensures that in the case that the container is empty, the data pointer access does not perform an errant memory access.

Could you re-test if this still happens?

**Comment by ChillerDragon** [comment]
Oh interesting. Clang hints are so good :)
Looking closer at the asan error I realized the error is in the vector code not in the mysql code. Until now I thought there was a null pointer passed to a mysql function. But it is actually the vector access that is causing the issue.

But what does ``data()`` return when the vector is empty?  Is it really safe to pass that to ``mysql_stmt_bind_param``? Maybe there is still UB but asan just does not catch it?

> Could you re-test if this still happens?

I could never consistently reproduce it. In ddnet-insta I run a query without parameters every round start. But it would on average happen a few times a day per game server. It now has not happend on any server since 3 days so I am pretty sure asan is pleased now (without this pr). I will wait a few more days just to be sure and then close the pr unless someone agrees with me that it is spooky to pass empty ``data()`` to ``mysql_stmt_bind_param()``.

**Comment by heinrich5991** [comment]
> Is it really safe to pass that to `mysql_stmt_bind_param`?

I would expect `mysql_stmt_bind_param` to not look at this parameter if there are no placeholders. But I haven't read their source code.

> I will wait a few more days just to be sure and then close the pr unless someone agrees with me that it is spooky to pass empty `data()` to `mysql_stmt_bind_param()`.

I'd also think that not making this change is better, unless it fixes an actual problem.

**Comment by ChillerDragon** [comment]
Still has not happend yet. So I guess we are safe. Closing.

**Pull request closed without merging**